File: test/kitchen/test/integration/system-probe-test/rspec/system-probe-test_spec.rb
require 'fileutils'
require 'kernel_out_spec_helper'
require 'open3'
require 'csv'
require 'rexml/document'

GOLANG_TEST_FAILURE = /FAIL:/

runtime_compiled_tests = Array.[](
  "pkg/network/tracer",
  "pkg/network/protocols/http",
)

co_re_tests = Array.[](
  "pkg/collector/corechecks/ebpf/probe"
)

print KernelOut.format(`cat /etc/os-release`)
print KernelOut.format(`uname -a`)

arch = `uname -m`.strip
release = `uname -r`.strip
osr = Hash[*CSV.read("/etc/os-release", col_sep: "=").flatten(1)]
platform = "#{osr["ID"]}-#{osr["VERSION_ID"]}"

##
## The main chef recipe (test\kitchen\site-cookbooks\dd-system-probe-check\recipes\default.rb)
## copies the necessary files (including the precompiled object files), and sets the mode to
## 0755, which causes the test to fail.  The object files are not being built during the
## test, anyway, so set them to the expected value
##
Dir.glob('/tmp/system-probe-tests/pkg/ebpf/bytecode/build/*.o').each do |f|
  FileUtils.chmod 0644, f, :verbose => true
end
Dir.glob('/tmp/system-probe-tests/pkg/ebpf/bytecode/build/co-re/*.o').each do |f|
  FileUtils.chmod 0644, f, :verbose => true
end

shared_examples "passes" do |bundle, env, filter|
  after :context do
    print KernelOut.format(`find "/tmp/pkgjson/#{bundle}" -maxdepth 1 -type f -path "*.json" -exec cat >"/tmp/testjson/#{bundle}.json" {} +`)
  end

  Dir.glob('/tmp/system-probe-tests/**/testsuite').each do |f|
    pkg = f.delete_prefix('/tmp/system-probe-tests/').delete_suffix('/testsuite')
    next unless filter.nil? or filter.include? pkg

    base_env = {
      "DD_SYSTEM_PROBE_BPF_DIR"=>"/tmp/system-probe-tests/pkg/ebpf/bytecode/build",
      "GOVERSION"=>"unknown"
    }
    junitfile = pkg.gsub("/","-") + ".xml"

    it "#{pkg} tests" do |ex|
      Dir.chdir(File.dirname(f)) do
        xmlpath = "/tmp/junit/#{bundle}/#{junitfile}"
        cmd = ["sudo", "-E",
          "/go/bin/gotestsum",
          "--format", "dots",
          "--junitfile", xmlpath,
          "--jsonfile", "/tmp/pkgjson/#{bundle}/#{pkg.gsub("/","-")}.json",
          "--raw-command", "--",
          "/go/bin/test2json", "-t", "-p", pkg, f, "-test.v", "-test.count=1"
        ]

        final_env = base_env.merge(env)
        Open3.popen2e(final_env, *cmd) do |_, output, wait_thr|
          output.each_line do |line|
            puts KernelOut.format(line.strip)
          end
        end

        xmldoc = REXML::Document.new(File.read(xmlpath))
        REXML::XPath.each(xmldoc, "//testsuites/testsuite/properties") do |props|
          props.add_element("property", { "name" => "dd_tags[test.bundle]", "value" => bundle })
          props.add_element("property", { "name" => "dd_tags[os.platform]", "value" => platform })
          props.add_element("property", { "name" => "dd_tags[os.architecture]", "value" => arch })
          props.add_element("property", { "name" => "dd_tags[os.version]", "value" => release })
        end
        File.open(xmlpath, "w") do |f|
          xmldoc.write(:output => f, :indent => 4)
        end
      end
    end
  end
end

describe "system-probe" do
  after :all do
    print KernelOut.format(`tar -C /tmp/junit -czf /tmp/junit.tar.gz .`)
    print KernelOut.format(`tar -C /tmp/testjson -czf /tmp/testjson.tar.gz .`)
  end

  context "prebuilt" do
    env = {
      "DD_ENABLE_RUNTIME_COMPILER"=>"false",
      "DD_ENABLE_CO_RE"=>"false"
    }
    include_examples "passes", "prebuilt", env
  end

  context "runtime compiled" do
    env = {
      "DD_ENABLE_RUNTIME_COMPILER"=>"true",
      "DD_ALLOW_PRECOMPILED_FALLBACK"=>"false",
      "DD_ENABLE_CO_RE"=>"false"
    }
    include_examples "passes", "runtime", env, runtime_compiled_tests
  end

  context "CO-RE" do
    env = {
      "DD_ENABLE_CO_RE"=>"true",
      "DD_ENABLE_RUNTIME_COMPILER"=>"false",
      "DD_ALLOW_RUNTIME_COMPILED_FALLBACK"=>"false"
    }
    include_examples "passes", "co-re", env, co_re_tests
  end
end

File: test/new-e2e/system-probe/system-probe-test_spec.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"io/fs"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
)

const (
	Testsuite   = "testsuite"
	TestDirRoot = "/opt/system-probe-tests"
	Sudo        = "sudo"
)

var BaseEnv = map[string]interface{}{
	"DD_SYSTEM_PROBE_BPF_DIR": filepath.Join("/opt/system-probe-tests", "pkg/ebpf/bytecode/build"),
}

type testConfig struct {
	bundle         string
	env            map[string]interface{}
	filterPackages filterPaths
}

type filterPaths struct {
	paths     []string
	inclusive bool
}

var skipPrebuiltTests = filterPaths{
	paths:     []string{"pkg/collector/corechecks/ebpf/probe"},
	inclusive: false,
}

var runtimeCompiledTests = filterPaths{
	paths: []string{
		"pkg/network/tracer",
		"pkg/network/protocols/http",
		"pkg/collector/corechecks/ebpf/probe",
	},
	inclusive: true,
}

var coreTests = filterPaths{
	paths: []string{
		"pkg/collector/corechecks/ebpf/probe",
		"pkg/network/protocols/http",
	},
	inclusive: true,
}

var fentryTests = filterPaths{
	paths:     skipPrebuiltTests.paths,
	inclusive: false,
}

func pathEmbedded(fullPath, embedded string) bool {
	normalized := fmt.Sprintf("/%s/",
		strings.TrimRight(
			strings.TrimLeft(embedded, "/"),
			"/",
		),
	)

	return strings.Contains(fullPath, normalized)
}

func glob(dir, filePattern string, filter filterPaths) ([]string, error) {
	var matches []string
	err := filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}

		present, err := regexp.Match(filePattern, []byte(d.Name()))
		if err != nil {
			return err
		}

		if d.IsDir() || !present {
			return nil
		}
		for _, p := range filter.paths {
			if pathEmbedded(path, p) && filter.inclusive {
				matches = append(matches, path)
			} else if !pathEmbedded(path, p) && !filter.inclusive {
				matches = append(matches, path)
			}
		}
		return nil
	})
	if err != nil {
		return []string{}, err
	}

	return matches, nil
}

func generatePacakgeName(file string) string {
	pkg := strings.TrimLeft(
		strings.TrimRight(
			strings.TrimPrefix(
				strings.TrimSuffix(file, Testsuite),
				TestDirRoot,
			),
			"/"),
		"/")

	return pkg
}

func buildCommandArgs(file, bundle string) []string {
	pkg := generatePacakgeName(file)
	junitfilePrefix := strings.ReplaceAll(pkg, "/", "-")
	xmlpath := filepath.Join(
		"/", "tmp", bundle,
		fmt.Sprintf("%s.xml", junitfilePrefix),
	)
	jsonpath := filepath.Join(
		"/", "tmp", bundle,
		fmt.Sprintf("%s.json", junitfilePrefix),
	)
	args := []string{
		"-E",
		"/go/bin/gotestsum",
		"--format", "dots",
		"--junitfile", xmlpath,
		"--jsonfile", jsonpath,
		"--raw-command", "--",
		"/go/bin/test2json", "-t", "-p", pkg, file, "-test.v", "-test.count=1",
	}

	return args
}

func mergeEnv(env ...map[string]interface{}) []string {
	var mergedEnv []string

	for _, e := range env {
		for key, element := range e {
			mergedEnv = append(mergedEnv, fmt.Sprintf("%s=%s", key, fmt.Sprint(element)))
		}
	}

	return mergedEnv
}

func runCommandAndStreamOutput(cmd *exec.Cmd, commandOutput io.Reader) error {
	go func() {
		scanner := bufio.NewScanner(commandOutput)
		for scanner.Scan() {
			_, _ = os.Stdout.Write([]byte(scanner.Text() + "\n"))
		}
	}()

	return cmd.Run()
}

func testPass(config testConfig) error {
	matches, err := glob(TestDirRoot, Testsuite, config.filterPackages)
	if err != nil {
		return err
	}

	for _, file := range matches {
		args := buildCommandArgs(file, config.bundle)
		fmt.Println(args)
		cmd := exec.Command(Sudo, args...)

		r, w := io.Pipe()
		cmd.Env = mergeEnv(config.env, BaseEnv)
		cmd.Dir = filepath.Dir(file)
		cmd.Stdout = w
		cmd.Stderr = w

		if err := runCommandAndStreamOutput(cmd, r); err != nil {
			return err
		}
	}

	return nil
}

func fixAssetPermissions() error {
	matches, err := glob(TestDirRoot, `.*\.o`, filterPaths{
		paths:     []string{"pkg/ebpf/bytecode/build"},
		inclusive: true,
	})
	if err != nil {
		return err
	}

	for _, file := range matches {
		if err := os.Chown(file, 0, 0); err != nil {
			return err
		}
	}

	return nil
}

func main() {
	if err := fixAssetPermissions(); err != nil {
		log.Fatal(err)
	}

	if err := testPass(testConfig{
		bundle: "prebuilt",
		env: map[string]interface{}{
			"DD_ENABLE_RUNTIME_COMPILER": false,
			"DD_ENABLE_CO_RE":            false,
		},
		filterPackages: skipPrebuiltTests,
	}); err != nil {
		log.Fatal(err)
	}
	if err := testPass(testConfig{
		bundle: "runtime",
		env: map[string]interface{}{
			"DD_ENABLE_RUNTIME_COMPILER":    true,
			"DD_ALLOW_PRECOMPILED_FALLBACK": false,
			"DD_ENABLE_CO_RE":               false,
		},
		filterPackages: runtimeCompiledTests,
	}); err != nil {
		log.Fatal(err)
	}
	if err := testPass(testConfig{
		bundle: "co-re",
		env: map[string]interface{}{
			"DD_ENABLE_CO_RE":                    true,
			"DD_ENABLE_RUNTIME_COMPILER":         false,
			"DD_ALLOW_RUNTIME_COMPILED_FALLBACK": false,
			"DD_ALLOW_PRECOMPILED_FALLBACK":      false,
		},
		filterPackages: coreTests,
	}); err != nil {
		log.Fatal(err)
	}
	if err := testPass(testConfig{
		bundle: "fentry",
		env: map[string]interface{}{
			"ECS_FARGATE":                   true,
			"DD_ENABLE_CO_RE":               true,
			"DD_ENABLE_RUNTIME_COMPILER":    false,
			"DD_ALLOW_PRECOMPILED_FALLBACK": false,
		},
		filterPackages: fentryTests,
	}); err != nil {
		log.Fatal(err)
	}
}
